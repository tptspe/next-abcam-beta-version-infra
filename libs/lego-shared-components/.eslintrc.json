{
  "extends": [
    "plugin:@nrwl/nx/react",
    "../../.eslintrc.json",
    "plugin:jsx-a11y/recommended",
    "plugin:cypress/recommended"
  ],
  "ignorePatterns": ["!**/*", "*.stories.*", "*.spec.*", ".storybook/*"],
  "plugins": ["no-relative-import-paths"],
  "rules": {
    "no-relative-import-paths/no-relative-import-paths": [
      "warn",
      { "allowSameFolder": true }
    ]
  },
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "parserOptions": {
        "project": ["libs/lego-shared-components/tsconfig.*?.json"]
      },
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    }
  ]
}
