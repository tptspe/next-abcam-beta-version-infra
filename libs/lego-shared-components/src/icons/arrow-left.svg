<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="none" viewBox="0 0 16 16">
    <path fill="#000" d="M4.116 8.667h9.217c.368 0 .667-.299.667-.667 0-.368-.299-.667-.667-.667H4.116l3.057-3.566c.24-.28.207-.7-.072-.94-.28-.24-.7-.207-.94.072l-4 4.667c-.215.25-.215.618 0 .868l4 4.667c.24.28.66.311.94.072.28-.24.311-.66.072-.94L4.116 8.667z"/>
</svg>
