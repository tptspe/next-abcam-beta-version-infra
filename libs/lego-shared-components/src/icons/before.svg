<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" fill="none" viewBox="0 0 20 20">
    <path fill="#fff" d="M7.5 12.988l7.744-7.744c.325-.325.853-.325 1.178 0 .326.326.326.853 0 1.179L8.09 14.756c-.325.325-.853.325-1.178 0l-3.334-3.333c-.325-.326-.325-.854 0-1.179.326-.325.853-.325 1.179 0L7.5 12.988z"/>
</svg>
