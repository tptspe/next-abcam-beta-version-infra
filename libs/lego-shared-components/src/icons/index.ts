export { ReactComponent as ChevronDown } from './chevron-down.svg'
export { ReactComponent as ChevronUp } from './chevron-up.svg'
export { ReactComponent as ArrowLeft } from './arrow-left.svg'
export { ReactComponent as Picture } from './picture.svg'
export { ReactComponent as Bin } from './bin.svg'
export { ReactComponent as Cart } from './cart.svg'
export { ReactComponent as Before } from './before.svg'
