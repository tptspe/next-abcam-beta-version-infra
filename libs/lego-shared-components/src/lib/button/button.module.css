.container {
  @apply inline-flex;
  @apply self-center;
  @apply items-center;
  @apply mx-1;
  @apply border;
  @apply max-w-max;
  @apply min-w-max;

  path {
    @apply fill-current;
  }

  .content {
    @apply block;
    @apply px-1.5;
    padding-bottom: 0.19rem;
  }

  &:hover {
    @apply no-underline;
  }

  &:focus {
    @apply outline-none;
    @apply shadow-interactiveElement;
  }

  &.disabled {
    @apply bg-grey90;
    @apply border-grey90;
    @apply text-white;
    @apply cursor-not-allowed;

    &:hover {
      @apply bg-grey90;
    }
  }

  &.iconButton {
    min-width: auto;
  }
}

/****** Variants ******/

.primary {
  @apply bg-buttonPrimary;
  @apply border-buttonPrimary;
  @apply text-white;

  &:hover {
    @apply bg-buttonPrimaryHover;
  }

  &:active {
    @apply bg-interactiveBlueActive;
  }
}

.secondary {
  @apply bg-grey35;
  @apply text-white;
  @apply border-grey35;

  &:hover {
    @apply bg-buttonSecondaryHover;
  }

  &:active {
    @apply bg-interactiveGreyDarkActive;
  }
}

.tertiary {
  @apply bg-white;
  @apply text-grey20;
  @apply border-backgroundLightbox;

  &:hover {
    @apply bg-interactiveGreyTransparentHover;
  }

  &:active {
    @apply bg-backgroundLightbox;
  }

  &.disabled {
    @apply bg-white;
    @apply text-grey80;
  }
}

.quaternary {
  @apply bg-buttonQuaternary;
  @apply border-buttonQuaternary;
  @apply text-grey20;

  &:hover {
    @apply bg-buttonQuaternaryHover;
  }
  &:active {
    @apply bg-interactiveGreyTransparentActive;
  }
}

.quinary {
  @apply border-0;
  @apply text-grey20;

  &:hover {
    @apply bg-interactiveGreyTransparentHover;
  }

  &:active {
    @apply bg-interactiveGreyTransparentActive;
  }

  &:disabled {
    @apply text-grey80;
    @apply bg-transparent;
  }
}

/****** Sizes ******/
.small {
  @apply px-2.5;
  @apply py-1.5;
  @apply rounded-sm;
  @apply text-ui-small;

  svg {
    width: 0.831rem;
    height: 0.831rem;
  }

  &.iconButton {
    @apply w-8;
    @apply h-8;
    @apply px-3 py-2;
    @apply flex;
    @apply flex-col;
    @apply rounded-full;
  }
}

.medium {
  @apply rounded-lg;
  @apply text-ui-medium;
  @apply px-5 py-2;

  svg {
    width: 1.044rem;
    height: 1.044rem;
  }

  &.iconButton {
    @apply w-10;
    @apply h-10;
    @apply flex;
    @apply flex-col;
    @apply rounded-full;
    padding: 0.625rem;
  }
}

.large {
  @apply px-6;
  @apply py-3.5;
  @apply rounded-lg;
  @apply text-ui-large;

  svg {
    @apply w-5;
    @apply h-5;
  }

  &.iconButton {
    @apply w-14;
    @apply h-14;
    @apply flex;
    @apply flex-col;
    @apply p-4;
    @apply rounded-full;
  }
}
