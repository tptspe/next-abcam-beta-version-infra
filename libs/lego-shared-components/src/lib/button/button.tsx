import React from 'react'
import cn from 'classnames'
import { useRouter } from 'next/router'
import type { FC, SyntheticEvent } from 'react'
import type { ButtonProps } from './button.type'

import styles from './button.module.css'

const Button: FC<ButtonProps> = ({
  as = 'button',
  children,
  className,
  bodyClassName,
  disabled,
  href = '',
  onClick,
  openInNewWindow = false,
  size = 'large',
  variant = 'primary',
  iconLeft,
  iconRight,
  iconButton = false,
  ...otherProps
}) => {
  const router = useRouter()

  const htmlClassName = cn(
    styles.container,
    styles[variant],
    styles[size],
    className,
    bodyClassName,
    { [styles.disabled]: disabled },
    { [styles.iconButton]: iconButton }
  )

  const handleAnchorClick = (e: SyntheticEvent) => {
    e.preventDefault()
    router.push(href)
  }

  const contentClassName = cn(styles.content, bodyClassName)
  if (as === 'button') {
    return (
      <button
        className={htmlClassName}
        disabled={disabled}
        onClick={onClick}
        type="button"
        {...otherProps}
      >
        {iconLeft && iconLeft}
        <span className={contentClassName}>{children}</span>
        {iconRight && iconRight}
      </button>
    )
  }
  return (
    <a
      className={htmlClassName}
      href={href}
      {...(!openInNewWindow && { onClick: handleAnchorClick })}
      {...(openInNewWindow && { rel: 'noreferrer', target: '_blank' })}
      {...otherProps}
    >
      {iconLeft && iconLeft}
      <span className={contentClassName}>{children}</span>
      {iconRight && iconRight}
    </a>
  )
}

export { Button }
