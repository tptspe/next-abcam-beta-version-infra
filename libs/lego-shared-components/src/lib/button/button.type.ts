import type { ReactNode } from 'react'

type ButtonProps = {
  as?: 'a' | 'button'
  children?: ReactNode
  className?: string
  bodyClassName?: string
  disabled?: boolean
  href?: string
  onClick?: () => void
  openInNewWindow?: boolean
  size?: 'small' | 'medium' | 'large'
  variant: 'primary' | 'secondary' | 'tertiary' | 'quaternary' | 'quinary'
  iconLeft?: ReactNode
  iconRight?: ReactNode
  iconButton?: boolean
}

export type { ButtonProps }
