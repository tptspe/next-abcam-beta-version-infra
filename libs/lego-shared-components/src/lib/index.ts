export { Button } from './button/button'
export { Select } from './select/select'
export { Dropdown } from './dropdown/dropdown'
