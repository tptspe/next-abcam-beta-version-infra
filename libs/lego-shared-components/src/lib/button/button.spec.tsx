import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import { useRouter } from 'next/router'
import { ButtonProps } from './button.type'
import { Button } from '@abcam-web/lego-shared-components/lib'
import { Picture } from '@abcam-web/lego-shared-components/icons'

jest.mock('next/router', () => ({ useRouter: jest.fn() }))

describe('The button', () => {
  describe('As a button', () => {
    it('Triggers the click callback on click', () => {
      const spy = jest.fn()
      render(
        <Button variant={'primary'} onClick={spy}>
          Hello
        </Button>
      )
      fireEvent.click(screen.getByText('Hello'))
      expect(spy.mock.calls.length).toBe(1)
    })

    it('Does not trigger the click callback if the disabled prop is passed', () => {
      const spy = jest.fn()
      render(
        <Button variant={'primary'} onClick={spy} disabled={true}>
          Hello
        </Button>
      )
      fireEvent.click(screen.getByText('Hello'))
      expect(spy.mock.calls.length).toBe(0)
    })
  })

  describe('As a link', () => {
    it('Handles internal links', () => {
      const pushMock = jest.fn()
      ;(useRouter as any).mockImplementation(() => {
        return {
          push: pushMock,
        }
      })
      const fakeHref = 'http://www.fake.com'
      render(
        <Button variant={'primary'} as={'a'} href={fakeHref}>
          Hello
        </Button>
      )
      fireEvent.click(screen.getByText('Hello'))
      expect(pushMock).toHaveBeenCalledTimes(1)
      fireEvent.click(screen.getByText('Hello'))
      expect(pushMock).toHaveBeenCalledTimes(2)
    })

    it('Handles external links', () => {
      const pushMock = jest.fn()
      ;(useRouter as any).mockImplementation(() => {
        return {
          push: pushMock,
        }
      })
      const fakeHref = 'http://www.fake.com'
      render(
        <Button variant={'primary'} as={'a'} href={fakeHref} openInNewWindow>
          Hello
        </Button>
      )
      fireEvent.click(screen.getByText('Hello'))
      expect(pushMock).not.toHaveBeenCalled()
    })
  })
  describe('Snapshots', () => {
    describe('Styles variant', () => {
      it('Accepts any combination of size and variant', () => {
        const sizes: ButtonProps['size'][] = ['small', 'medium', 'large']
        const variants: ButtonProps['variant'][] = [
          'primary',
          'secondary',
          'tertiary',
          'quaternary',
        ]

        sizes.forEach((size) => {
          variants.forEach((variant) => {
            const { container } = render(
              <Button variant={variant} size={size}>
                Hello
              </Button>
            )
            expect(container.firstChild).toMatchSnapshot()
          })
        })
      })
    })
    describe('As a button', () => {
      it('Matches snapshot', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button variant={'primary'} as={'button'} onClick={spy}>
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a left icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'button'}
            onClick={spy}
            iconLeft={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a right icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'button'}
            onClick={spy}
            iconRight={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a left and right icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'button'}
            onClick={spy}
            iconRight={<Picture />}
            iconLeft={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Shows a iconButton', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button variant={'primary'} as={'button'} onClick={spy} iconButton>
            <Picture />
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts external classes', () => {
        const spy = jest.fn()
        const fakeClass = 'myFakeClass'
        const { container } = render(
          <Button
            variant={'primary'}
            as={'button'}
            onClick={spy}
            className={fakeClass}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })
    })

    describe('As a link', () => {
      it('Matches snapshot', () => {
        const { container } = render(
          <Button variant={'primary'} as={'a'}>
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a left icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'a'}
            onClick={spy}
            iconLeft={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a right icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'a'}
            onClick={spy}
            iconRight={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts a left and right icon', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button
            variant={'primary'}
            as={'a'}
            onClick={spy}
            iconRight={<Picture />}
            iconLeft={<Picture />}
          >
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Shows a iconButton', () => {
        const spy = jest.fn()
        const { container } = render(
          <Button variant={'primary'} as={'a'} onClick={spy} iconButton>
            <Picture />
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })

      it('Accepts external classes', () => {
        const spy = jest.fn()
        const fakeClass = 'myFakeClass'
        const { container } = render(
          <Button variant={'primary'} as={'a'} className={fakeClass}>
            Hello
          </Button>
        )
        expect(container.firstChild).toMatchSnapshot()
      })
    })
  })
})
