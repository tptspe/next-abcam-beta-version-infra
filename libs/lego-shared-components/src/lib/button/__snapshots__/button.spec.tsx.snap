// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The button Snapshots As a button Accepts a left and right icon 1`] = `
<button
  class="container primary large"
  type="button"
>
  <svg />
  <span
    class="content"
  >
    Hello
  </span>
  <svg />
</button>
`;

exports[`The button Snapshots As a button Accepts a left icon 1`] = `
<button
  class="container primary large"
  type="button"
>
  <svg />
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots As a button Accepts a right icon 1`] = `
<button
  class="container primary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
  <svg />
</button>
`;

exports[`The button Snapshots As a button Accepts external classes 1`] = `
<button
  class="container primary large myFakeClass"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots As a button Matches snapshot 1`] = `
<button
  class="container primary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots As a button Shows a iconButton 1`] = `
<button
  class="container primary large iconButton"
  type="button"
>
  <span
    class="content"
  >
    <svg />
  </span>
</button>
`;

exports[`The button Snapshots As a link Accepts a left and right icon 1`] = `
<a
  class="container primary large"
  href=""
>
  <svg />
  <span
    class="content"
  >
    Hello
  </span>
  <svg />
</a>
`;

exports[`The button Snapshots As a link Accepts a left icon 1`] = `
<a
  class="container primary large"
  href=""
>
  <svg />
  <span
    class="content"
  >
    Hello
  </span>
</a>
`;

exports[`The button Snapshots As a link Accepts a right icon 1`] = `
<a
  class="container primary large"
  href=""
>
  <span
    class="content"
  >
    Hello
  </span>
  <svg />
</a>
`;

exports[`The button Snapshots As a link Accepts external classes 1`] = `
<a
  class="container primary large myFakeClass"
  href=""
>
  <span
    class="content"
  >
    Hello
  </span>
</a>
`;

exports[`The button Snapshots As a link Matches snapshot 1`] = `
<a
  class="container primary large"
  href=""
>
  <span
    class="content"
  >
    Hello
  </span>
</a>
`;

exports[`The button Snapshots As a link Shows a iconButton 1`] = `
<a
  class="container primary large iconButton"
  href=""
>
  <span
    class="content"
  >
    <svg />
  </span>
</a>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 1`] = `
<button
  class="container primary small"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 2`] = `
<button
  class="container secondary small"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 3`] = `
<button
  class="container tertiary small"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 4`] = `
<button
  class="container quaternary small"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 5`] = `
<button
  class="container primary medium"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 6`] = `
<button
  class="container secondary medium"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 7`] = `
<button
  class="container tertiary medium"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 8`] = `
<button
  class="container quaternary medium"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 9`] = `
<button
  class="container primary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 10`] = `
<button
  class="container secondary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 11`] = `
<button
  class="container tertiary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;

exports[`The button Snapshots Styles variant Accepts any combination of size and variant 12`] = `
<button
  class="container quaternary large"
  type="button"
>
  <span
    class="content"
  >
    Hello
  </span>
</button>
`;
